rust/worker/src/storage/s3.rs

```diff
@@ -156,55 +160,163 @@ impl S3Storage {
     }
 
     pub(crate) async fn put_bytes(&self, key: &str, bytes: Vec<u8>) -> Result<(), S3PutError> {
-        let bytestream = ByteStream::from(bytes);
-        self.put_bytestream(key, bytestream).await
-    }
-
-    pub(crate) async fn put_file(&self, key: &str, path: &str) -> Result<(), S3PutError> {
-        let bytestream = ByteStream::from_path(path).await;
-        match bytestream {
-            Ok(bytestream) => return self.put_bytestream(key, bytestream).await,
-            Err(e) => {
-                return Err(S3PutError::S3PutError(e.to_string()));
+        let upload_id = match self
+            .client
+            .create_multipart_upload()
+            .bucket(&self.bucket)
+            .key(key)
+            .send()
+            .await
+            .map_err(|err| S3PutError::S3PutError(err.to_string()))?
+            .upload_id
+        {
+            Some(upload_id) => upload_id,
+            None => {
+                return Err(S3PutError::S3PutError(
+                    "Multipart upload creation response missing upload ID".to_string(),
+                ));
             }
+        };
+
+        let mut upload_parts = Vec::new();
+        for (part_number, offset, length) in self.part_number_offset_length_iter(bytes.len() as u64)
+        {
+            let stream =
+                ByteStream::from(bytes[offset as usize..(offset + length) as usize].to_vec());
```

Comment:
this to_vec will result in the underlying data being copied?

Reply:
yes, I should have noted this--I didn't see a way to directly create a byte stream from a slice, but will double check

Reply:
turned out to be much more straightforward than I was thinking
a38b817a